src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,1074 @@
+    init_req->set_version(0);
+    init_req->mutable_tsx_data()->CopyFrom(tsx_data);
+    return init_req;
+  }
+
+  void Signer::step_init_ack(std::shared_ptr<const messages::monero::MoneroTransactionInitAck> ack){
+    m_ct.in_memory = ack->in_memory();
+    if (ack->has_rsig_data()){
+      m_ct.rsig_param = std::make_shared<MoneroRsigData>(ack->rsig_data());
+    }
+
+    assign_from_repeatable(&(m_ct.tx_out_entr_hmacs), ack->hmacs().begin(), ack->hmacs().end());
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionSetInputRequest> Signer::step_set_input(size_t idx){
+    m_ct.cur_input_idx = idx;
+    auto res = std::make_shared<messages::monero::MoneroTransactionSetInputRequest>();
+    translate_src_entry(res->mutable_src_entr(), &(cur_tx().sources[idx]));
+    return res;
+  }
+
+  void Signer::step_set_input_ack(std::shared_ptr<const messages::monero::MoneroTransactionSetInputAck> ack){
+    auto & vini_str = ack->vini();
+
+    cryptonote::txin_v vini;
+    if (!cn_deserialize(vini_str.data(), vini_str.size(), vini)){
+      throw exc::ProtocolException("Cannot deserialize vin[i]");
+    }
+
+    m_ct.tx.vin.emplace_back(vini);
+    m_ct.tx_in_hmacs.push_back(ack->vini_hmac());
+    m_ct.pseudo_outs.push_back(ack->pseudo_out());
+    m_ct.pseudo_outs_hmac.push_back(ack->pseudo_out_hmac());
+    m_ct.alphas.push_back(ack->alpha_enc());
+    m_ct.spend_encs.push_back(ack->spend_enc());
+  }
+
+  void Signer::sort_ki(){
+    m_ct.source_permutation.clear();
+    for (size_t n = 0; n < cur_tx().sources.size(); ++n){
+      m_ct.source_permutation.push_back(n);
+    }
+
+    std::sort(m_ct.source_permutation.begin(), m_ct.source_permutation.end(), [&](const size_t i0, const size_t i1) {
+      const cryptonote::txin_to_key &tk0 = boost::get<cryptonote::txin_to_key>(m_ct.tx.vin[i0]);
+      const cryptonote::txin_to_key &tk1 = boost::get<cryptonote::txin_to_key>(m_ct.tx.vin[i1]);
+      return memcmp(&tk0.k_image, &tk1.k_image, sizeof(tk0.k_image)) > 0;
+    });
+
+    tools::apply_permutation(m_ct.source_permutation, [&](size_t i0, size_t i1){
+      std::swap(m_ct.tx.vin[i0], m_ct.tx.vin[i1]);
+      std::swap(m_ct.tx_in_hmacs[i0], m_ct.tx_in_hmacs[i1]);
+      std::swap(m_ct.pseudo_outs[i0], m_ct.pseudo_outs[i1]);
+      std::swap(m_ct.pseudo_outs_hmac[i0], m_ct.pseudo_outs_hmac[i1]);
+      std::swap(m_ct.alphas[i0], m_ct.alphas[i1]);
+      std::swap(m_ct.spend_encs[i0], m_ct.spend_encs[i1]);
+      std::swap(m_ct.tx_data.sources[i0], m_ct.tx_data.sources[i1]);
+    });
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionInputsPermutationRequest> Signer::step_permutation(){
+    sort_ki();
+
+    if (in_memory()){
+      return nullptr;
+    }
+
+    auto res = std::make_shared<messages::monero::MoneroTransactionInputsPermutationRequest>();
+    assign_to_repeatable(res->mutable_perm(), m_ct.source_permutation.begin(), m_ct.source_permutation.end());
+
+    return res;
+  }
```

Comment:
Why do I get the feeling that this is mirroring a lot of existing code and will break when it changes... Is there no way to avoid this ?

Reply:
This probably got shifted after the rebase. Code deduplication will be done in the separate Lite PR. Lite protocol is now removed from this PR.

Reply:
Why not do it in one go ? That should also make review simpler at first approximation, since there's no review duplication either.

Reply:
A bit more on that is in this comment https://github.com/monero-project/monero/pull/4241#issuecomment-425991674

Currently, the Lite support is not integrated in the Trezor either so it was probably too ambitious to do it in one PR. Lite protocol is quite disjoint from the "cold wallet" signing protocol and properly implementing would require refactoring of the Ledger code also, which could double review effort for this PR (refactoring was not yet performed).